### PR TITLE
Print / store hashes / intermediate hashes

### PR DESCRIPTION
**Pull request by maurolacy**

### CODE REVIEW

File: optimize.sh [hunk 1]
```diff
@@ -12,6 +12,7 @@ echo "Info: sccache stats before build"
 sccache -s
 
 mkdir -p artifacts
+true >artifacts/checksums_intermediate.txt
```
Comment (webmaster128) [inline optimize.sh:15]: What is this supposed to do? Why not

```suggestion
touch artifacts/checksums_intermediate.txt
```

Comment (maurolacy) [reply optimize.sh]: No, this is not the same. Committed in haste. Will revert. What I want is to make sure the file doesn't exist / is empty.

maybe `rm -f` is clearer?

Comment (webmaster128) [reply optimize.sh]: `echo -n "" > artifacts/checksums_intermediate.txt` would be empty+existing
`rm -f` is you want non-existent

Comment (maurolacy) [reply optimize.sh]: Yes. I wasn't sure `tee -a` would work without a file, that's why i used that idiom.

Comment (maurolacy) [reply optimize.sh]: I really like `$ >file` idiom, by the way. It means: "Doing nothing produces nothing. And, redirecting nothing to a file, produces an empty file."

**Comment by maurolacy** [comment]
Here are the hashes on my Intel machine, for reference:
```
$ cat artifacts-amd64/checksums_intermediate.txt 
d377c8aeca37c1f3d3badd47639e73803ebfe476e95dead1ff6043c603d8028c  ./contracts/burner//target/wasm32-unknown-unknown/release/burner.wasm
dc08ab1132cd014357754da1848b4510dbb91676dff61ce8f821a536511f8655  ./contracts/crypto-verify//target/wasm32-unknown-unknown/release/crypto_verify.wasm
44c355dd263ba16849562a7b63e254db10ccce31cec66a4fa99c3e221d1b8f63  ./contracts/floaty//target/wasm32-unknown-unknown/release/floaty.wasm
edf01e678718311fe1e05030b7596d81cbcac46f8c1c2f5156cdf07fbb40b8ce  ./contracts/hackatom//target/wasm32-unknown-unknown/release/hackatom.wasm
726c0e05a42f090aa3014fa5d574b8d2b66273c025ff68838624ea26c6dc56df  ./contracts/ibc-reflect//target/wasm32-unknown-unknown/release/ibc_reflect.wasm
229f5c930060cfe5bffeb145bc722f55f72b97472cda58814e93fec94bd99e32  ./contracts/ibc-reflect-send//target/wasm32-unknown-unknown/release/ibc_reflect_send.wasm
5afaed3d7ce8c44e0c4adfa7d7ccb39a1608d216a1699843c1ec58d8d181b3b8  ./contracts/queue//target/wasm32-unknown-unknown/release/queue.wasm
713e78c48a410902c5eafb1777c7e8e6bd89abccb25d5df29380fc9eb6af5663  ./contracts/reflect//target/wasm32-unknown-unknown/release/reflect.wasm
816910146faefbc0be51c8fb9790060b4d0066ee97aec10c0d188249b53ad009  ./contracts/staking//target/wasm32-unknown-unknown/release/staking.wasm
```

```$ cat artifacts-amd64/checksums.txt 
f944023741e5982150ffd97f37201a047baafb451db4f86c438ae00fd1506d60  burner.wasm
f9c09829039c33111c3daf575bfabbecd7afde592f4eb0b2d745926fb9ae0d13  crypto_verify.wasm
585c6eff15526c035c9dbd8a7bce87005c84874dcb8fb7a841a2508362b1d95f  floaty.wasm
f8f4cab7fd5046eb75b964a00cab06686f9fbc15c17e4c4ac2ecd10678ac9546  hackatom.wasm
409c33790bbc41902cc958383f0fcac61a48b85ff394a89910795055ed552da2  ibc_reflect.wasm
3259f5889a3142e1c786b10a72e8f2b3f51e8910938b87649c9f8f1fda0feab8  ibc_reflect_send.wasm
ad0e69dce7231c95b54a1506d4ebb72c4f0849bd9a6b7695fb04fd6e58bbd99b  queue.wasm
fa5474312479a279eaec46d17cd27ef8d79724d16ff8251c26ce9b22e903bece  reflect.wasm
d88b4fd7726ee51a118ab813f7dd7f697e441dce03afceee607cabea9ec57481  staking.wasm
$ 
```

**Comment by harryscholes** [comment]
Can you link me to the repo for these contracts and I'll try running on native arm64

**Comment by maurolacy** [comment]
> Can you link me to the repo for these contracts and I'll try running on native arm64

https://github.com/CosmWasm/cosmwasm. Just switch to the `v0.16.2` tag, and run the latest `rust-optimizer` on the root.

**Comment by maurolacy** [comment]
> Can you link me to the repo for these contracts and I'll try running on native arm64

https://github.com/CosmWasm/cosmwasm. Just switch to the `v0.16.2` tag, and run the latest `rust-optimizer` on the root.

**Comment by harryscholes** [comment]
I'm getting a load of compiler errors for memmap2 on v0.16.2. I take it you're not seeing these

```
...
   Compiling memmap2 v0.2.3
error[E0061]: this function takes 3 arguments but 4 arguments were supplied
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/lib.rs:229:9
    |
229 |         MmapInner::map(self.get_len(file)?, file, self.offset, self.populate)
    |         ^^^^^^^^^^^^^^ -------------------  ----  -----------  ------------- supplied 4 arguments
    |         |
    |         expected 3 arguments
    |
note: associated function defined here
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/stub.rs:18:12
    |
18  |     pub fn map(_: usize, _: &File, _: u64) -> io::Result<MmapInner> {
    |            ^^^ --------  --------  ------

   Compiling getrandom v0.1.16
error[E0061]: this function takes 3 arguments but 4 arguments were supplied
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/lib.rs:240:9
    |
240 |         MmapInner::map_exec(self.get_len(file)?, file, self.offset, self.populate)
    |         ^^^^^^^^^^^^^^^^^^^ -------------------  ----  -----------  ------------- supplied 4 arguments
    |         |
    |         expected 3 arguments
    |
note: associated function defined here
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/stub.rs:22:12
    |
22  |     pub fn map_exec(_: usize, _: &File, _: u64) -> io::Result<MmapInner> {
    |            ^^^^^^^^ --------  --------  ------

error[E0061]: this function takes 3 arguments but 4 arguments were supplied
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/lib.rs:278:9
    |
278 |         MmapInner::map_mut(self.get_len(file)?, file, self.offset, self.populate)
    |         ^^^^^^^^^^^^^^^^^^ -------------------  ----  -----------  ------------- supplied 4 arguments
    |         |
    |         expected 3 arguments
    |
note: associated function defined here
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/stub.rs:26:12
    |
26  |     pub fn map_mut(_: usize, _: &File, _: u64) -> io::Result<MmapInner> {
    |            ^^^^^^^ --------  --------  ------

error[E0061]: this function takes 3 arguments but 4 arguments were supplied
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/lib.rs:307:9
    |
307 |         MmapInner::map_copy(self.get_len(file)?, file, self.offset, self.populate)
    |         ^^^^^^^^^^^^^^^^^^^ -------------------  ----  -----------  ------------- supplied 4 arguments
    |         |
    |         expected 3 arguments
    |
note: associated function defined here
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/stub.rs:30:12
    |
30  |     pub fn map_copy(_: usize, _: &File, _: u64) -> io::Result<MmapInner> {
    |            ^^^^^^^^ --------  --------  ------

error[E0061]: this function takes 3 arguments but 4 arguments were supplied
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/lib.rs:340:9
    |
340 |         MmapInner::map_copy_read_only(self.get_len(file)?, file, self.offset, self.populate)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ -------------------  ----  -----------  ------------- supplied 4 arguments
    |         |
    |         expected 3 arguments
    |
note: associated function defined here
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/stub.rs:34:12
    |
34  |     pub fn map_copy_read_only(_: usize, _: &File, _: u64) -> io::Result<MmapInner> {
    |            ^^^^^^^^^^^^^^^^^^ --------  --------  ------

error[E0061]: this function takes 3 arguments but 4 arguments were supplied
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/lib.rs:363:9
    |
363 |         MmapInner::map_mut(self.get_len(file)?, file, self.offset, self.populate)
    |         ^^^^^^^^^^^^^^^^^^ -------------------  ----  -----------  ------------- supplied 4 arguments
    |         |
    |         expected 3 arguments
    |
note: associated function defined here
   --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/memmap2-0.2.3/src/stub.rs:26:12
    |
26  |     pub fn map_mut(_: usize, _: &File, _: u64) -> io::Result<MmapInner> {
    |            ^^^^^^^ --------  --------  ------

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0061`.
error: could not compile `memmap2`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...

error: build failed
```

**Comment by maurolacy** [comment]
Not sure example why that happens. In the mean time, try with this:
```
docker run --rm -v "$(pwd)":/code --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target  maurolacy/rust-optimizer:0.12.2 contracts/*/
```

**Comment by harryscholes** [comment]
I got the same results as your emulated arm64 results

```
❯ cat artifacts/checksums_intermediate.txt
00a07c01e5deaa4cbe4b359424c40c14c0451abf399df7b627b5ffccd685e2bf  contracts/burner//target/wasm32-unknown-unknown/release/burner.wasm
4fd5a41ee350a5b7039c9fb1ee83beffdc31bca98815e396675d1370c164a51c  contracts/crypto-verify//target/wasm32-unknown-unknown/release/crypto_verify.wasm
7475374a3935fffd350deab21d870bc82c09c06d6967f14adf206e8e3eff424a  contracts/floaty//target/wasm32-unknown-unknown/release/floaty.wasm
8192c034a4599b9fcc0ce2d5f00edb857145de00449142d1656fcd1bd2846b92  contracts/hackatom//target/wasm32-unknown-unknown/release/hackatom.wasm
97ad2777aa42829d73dac6c0d8c1bd15e4ffb0e4e21ee67a55244fc385c26046  contracts/ibc-reflect-send//target/wasm32-unknown-unknown/release/ibc_reflect_send.wasm
13701132d236a7fa1b5b6fe1a6d06b6ca6c28008d3029eb45d161c1f4afb2ca0  contracts/ibc-reflect//target/wasm32-unknown-unknown/release/ibc_reflect.wasm
3ce0bffe2be9bf3b3c5ac0691ac5faef839f180a2a75b89b752565cbcaf84a25  contracts/queue//target/wasm32-unknown-unknown/release/queue.wasm
b0579203caaa28d54c7eb0559b2ea1d82893d27528aec3d0c9b333066d71388b  contracts/reflect//target/wasm32-unknown-unknown/release/reflect.wasm
77751c697279cc35adec424265a7ecd19fbb698c65accfdd40eb76535155fd5e  contracts/staking//target/wasm32-unknown-unknown/release/staking.wasm
❯ cat artifacts/checksums.txt
ffd6e3824dc4ce7250416b99eb7cb51996783350d87da4910a210122fc11bf94  burner.wasm
850f1a95d8426c77f6e5e8ab1523d9c89a6365a95c43ae132dc56626ef4b92eb  crypto_verify.wasm
3ed81f2506e7369abf5f55cd415c95b2a0bf858f30facd9c82bee45fb34abeab  floaty.wasm
5780e953070c5207a9c3ba3f310502cb0efeacb79b1b4b4c3a7a66091209d91f  hackatom.wasm
551ad6c60a5d69a89481b014ebaf6a1dc86a89b992a425816618dc8906bb7b98  ibc_reflect.wasm
7019c1f0d060b4f1887f5d3f70ff547371f387f9ea9fbf231723299448171be6  ibc_reflect_send.wasm
368d944f8dd90422bbcf88986f91fabf822bcb6b801512f9c161e796acddf61a  queue.wasm
1be355973ae5938077973e8dcb62b719c45280f5d8cbcb75fcedf4c168d4541a  reflect.wasm
624b2cd85b70568023f7b29ab9ceb92d51460f68d73f8806af5b6096a4aab955  staking.wasm
```

**Comment by maurolacy** [comment]
Thanks. Merging this.

**Review by webmaster128 — APPROVED** [review]
🥇

**Pull request merged**